lib/proxy/launcher.rb

```diff
@@ -79,8 +79,16 @@ def launch
 
       ::Proxy::PluginInitializer.new(plugins).initialize_plugins
 
-      require 'proxy/launcher/webrick'
-      launcher = ::Launcher::Webrick.new(self)
+      case settings.http_server_type
+      when 'webrick'
+        require 'proxy/launcher/webrick'
+        launcher = ::Launcher::Webrick.new(self)
+      when 'puma'
+        require 'proxy/launcher/puma'
+        launcher = ::Launcher::Puma.new(self)
```

Comment:
Why is this needed? Aren't we actually switching from using our own startup binary to calling "puma" directly?

Reply:
This was a bit of my confusion as well trying to understand the layers and I think @lzap worded the question better than I was going to. I attempted to read up on the Rack pattern and it's uses in places and apply it to here. I'm going to attempt to write out what I think I understand (and am probably wrong about in some way):

```
Puma  -> Rack Middleware -> Smart Proxy Plugins
```

Where Puma serves as the application server handling spawning processes and threads to handle incoming requests.
Rack middleware (Proxy::App) handles mounting each Smart Proxy plugin as a Rack app as an http, https or both endpoint.

Reply:
> Why is this needed? Aren't we actually switching from using our own startup binary to calling "puma" directly?

Both are valid. The Puma launcher keeps everything compatible from a packaging perspective: no need to change all config files and service files.

> `Puma -> Rack Middleware -> Smart Proxy Plugins`

This is slightly incorrect. More correct would be:

Application server -> Rack application

Passenger would also be a valid application server since there is a `config.ru` file. Any Rack application can also apply middleware (we don't do this today).

If you note, previously it also built a Rack application here:

https://github.com/theforeman/smart-proxy/blob/88fbc8e67d665e2c3b19acb53b31ff30acf078b7/lib/launcher.rb#L47-L49

And the HTTPS one here:

https://github.com/theforeman/smart-proxy/blob/88fbc8e67d665e2c3b19acb53b31ff30acf078b7/lib/launcher.rb#L69-L71

So the solution I chose was to write a third Rack application that embeds boths those and switches it based on the received protocol.

Note that a Rack application is defined as a class that has a `call(env)` method. Env is a simple Ruby hash. To get a request, you can call `Rack::Request.new(env)`. It's surprisingly simple.

https://github.com/rack/rack/blob/master/lib/rack/lobster.rb is the example they ship.